osu.Game/Rulesets/UI/Playfield.cs

```diff
@@ -473,7 +473,7 @@ public DrawableSamplePool(ISampleInfo sampleInfo, int initialSize, int? maximumS
 
         private void onNewResult(DrawableHitObject drawable, JudgementResult result)
         {
-            Debug.Assert(result != null && drawable.Entry?.Result == result && result.RawTime != null);
+            Debug.Assert(result != null && drawable.Entry?.Result == result && result.RawTime != null && result.GameplayRate != 0.0);
```

Comment:
This zero check is worrying? What's it for?

Reply:
`RawTime` is populated in the same place as the new `GameplayRate` field, so I added this to match, though I couldn't actually think of a circumstance where `RawTime` would be null either, other than after `Reset`. Are there are other circumstances I'm missing? Should `GameplayRate` be reset in that method as well?

Reply:
Okay, sure. But then I'd argue that `GameplayRate` should be nullable rather than default to zero, and you'd assert that it's not null.

Reply:
`CalculateUnstableRate` in `HitEventExtensions` relies on there being a rate for every hit event, so either the fallback would have to go there or it could return null if any of the `GameplayRate`s is null. I'm not sure if I'm approaching this correctly, I don't really even understand the circumstances where the `IGameplayClock` could be null.

Reply:
The asserts are just for safety purposes. My argument is that if something is to ever go wrong here in the future and the gameplay clock _were_ to go missing at some point, the rate being `null` is a better indicator of what's going wrong than it being zero.

Usages should also assert non-null.

Reply:
Okay, so while this works ingame this causes tests to fail because there's no `IGameplayClock`. Initially the fallback was in place because when I was using DI it yielded null in tests, which I assumed was due to nothing caching that type, but I guess it's deeper than that. Is that sufficient justification for including the fallback?

Reply:
In that case I would say it's fine to add back the fallback to 1 where it was, but also remove the assertion because it has no functional purpose other than confusing whoever is reading it.